modules/table/src/integrationTest/java/org/apache/ignite/internal/table/ItTxResourcesVacuumTest.java

```diff
@@ -0,0 +1,798 @@
+        String sql = "create table " + TABLE_NAME + " (key bigint primary key, val varchar(20)) with primary_zone='TEST_ZONE'";
+
+        cluster.doInSession(0, session -> {
+            executeUpdate(zoneSql, session);
+            executeUpdate(sql, session);
+        });
+
+        txStateStorageExecutor = Executors.newSingleThreadExecutor(IgniteThreadFactory.create("test", "tx-state-storage-test-pool", log,
+                ThreadOperation.STORAGE_READ));
+    }
+
+    @Override
+    @AfterEach
+    public void tearDown() {
+        shutdownAndAwaitTermination(txStateStorageExecutor, 10, TimeUnit.SECONDS);
+
+        super.tearDown();
+    }
+
+    @Override
+    protected void customizeInitParameters(InitParametersBuilder builder) {
+        super.customizeInitParameters(builder);
+
+        builder.clusterConfiguration("{"
+                + "  transaction: {"
+                + "      txnResourceTtl: 1"
+                + "  },"
+                + "  replication: {"
+                + "      rpcTimeout: 30000"
+                + "  },"
+                + "}");
+    }
+
+    /**
+     * Returns node bootstrap config template.
+     *
+     * @return Node bootstrap config template.
+     */
+    @Override
+    protected String getNodeBootstrapConfigTemplate() {
+        return NODE_BOOTSTRAP_CFG_TEMPLATE;
+    }
+
+    /**
+     * Simple vacuum test, checking also that PENDING and FINISHING states are not removed.
+     *
+     * <ul>
+     *     <li>Run a transaction;</li>
+     *     <li>Insert a value;</li>
+     *     <li>Wait 3 seconds;</li>
+     *     <li>Check that the volatile PENDING state of the transaction is preserved;</li>
+     *     <li>Block {@link TxFinishReplicaRequest};</li>
+     *     <li>Start the tx commit;</li>
+     *     <li>While the state is FINISHING, wait 3 seconds;</li>
+     *     <li>Check that the volatile state of the transaction is preserved;</li>
+     *     <li>Unblock {@link TxFinishReplicaRequest};</li>
+     *     <li>Check that both volatile and persistent state is vacuumized;</li>
+     *     <li>Check that the committed value is correct.</li>
+     * </ul>
+     */
+    @Test
+    public void testVacuum() throws InterruptedException {
+        IgniteImpl node = anyNode();
+
+        RecordView<Tuple> view = node.tables().table(TABLE_NAME).recordView();
+
+        // Put some value into the table.
+        Transaction tx = node.transactions().begin();
+        UUID txId = txId(tx);
+
+        log.info("Test: Loading the data [tx={}].", txId);
+
+        Tuple tuple = findTupleToBeHostedOnNode(node, TABLE_NAME, tx, INITIAL_TUPLE, NEXT_TUPLE, true);
+
+        int partId = partitionIdForTuple(node, TABLE_NAME, tuple, tx);
+
+        Set<String> nodes = partitionAssignment(node, new TablePartitionId(tableId(node, TABLE_NAME), partId));
+
+        view.upsert(tx, tuple);
+
+        // Check that the volatile PENDING state of the transaction is preserved.
+        Thread.sleep(3000);
```

Comment:
Let's adjust the test in order to have waitForCondition instead of Thread.sleep. In case of no-action awaiting, let's just check that vacuumization happened but pending still there.

Reply:
removed sleeps